HubSpot.NET/Api/OAuth/HubSpotOAuthApi.cs

```diff
@@ -0,0 +1,155 @@
+        {
+            RequestRefreshTokenHubSpotModel model = new RequestRefreshTokenHubSpotModel()
+            {
+                ClientId = ClientId,
+                ClientSecret = _clientSecret,
+                RedirectUri = redirectUri,
+                RefreshToken = token.RefreshToken
+            };
+
+            HubSpotToken refreshToken = InitiateRequest(model, _client.BasePath);
+            _client.UpdateToken(refreshToken);
+            return refreshToken;
+        }
+
+        public void UpdateCredentials(string id, string secret)
+        {
+            ClientId = id;
+            _clientSecret = secret;
+        }
+
+        private HubSpotToken InitiateRequest<K>(K model, string basePath, params OAuthScopes[] scopes)
+        {
+            RestClient client = new RestClient(basePath);
+
+            StringBuilder builder = new StringBuilder();
+            foreach (OAuthScopes scope in scopes)
+            {
+                if (builder.Length == 0)
+                {
+                    builder.Append($"{OAuthScopeNameConversions[scope]}");
+                }
+                else
+                {
+                    builder.Append($"%20{OAuthScopeNameConversions[scope]}");
+                }
+            }
+
+            RestRequest request = new RestRequest(MidRoute)
+            {
+                JsonSerializer = new FakeSerializer()
+            };
+
+            Dictionary<string, string> jsonPreStringPairs = JsonConvert.DeserializeObject<Dictionary<string, string>>(JsonConvert.SerializeObject(model));
+
+            StringBuilder bodyBuilder = new StringBuilder();
+            foreach(KeyValuePair<string,string> pair in jsonPreStringPairs)
+            {
+                if (bodyBuilder.Length > 0)
+                { bodyBuilder.Append("&"); }
+
+                bodyBuilder.Append($"{pair.Key}={pair.Value}");
+            }
+
+            request.AddJsonBody(bodyBuilder.ToString());
+            request.AddHeader("ContentType", "application/x-www-form-urlencoded");
+
+            if (builder.Length > 0)
+                request.AddQueryParameter("scope", builder.ToString());
+
+            IRestResponse<HubSpotToken> serverReponse = client.Post<HubSpotToken>(request);
+
+            if (serverReponse.ResponseStatus != ResponseStatus.Completed)
+            {
+                throw new HubSpotException("Server did not respond to authorization request. Content: " + serverReponse.Content, new HubSpotError(serverReponse.StatusCode, serverReponse.Content), serverReponse.Content);
+            }
+
+            if (serverReponse.StatusCode == System.Net.HttpStatusCode.BadRequest)
+            {
+                throw new HubSpotException("Error generating authentication token.", JsonConvert.DeserializeObject<HubSpotError>(serverReponse.Content), serverReponse.Content);
+            }
+
+            return serverReponse.Data;
+        }
+    }
+
+    internal class FakeSerializer : ISerializer
```

Comment:
What's this used for?

Reply:
Good question. I remember it having to be a workaround for something with RestSharp. Vaguely though.